src/vdaf/mastic.rs

```diff
@@ -379,32 +391,500 @@ where
+                    Some(query_share) => query_share.encoded_len()?,
+                    None => 0,
+                },
+        )
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareShare<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match (&prep_state.szk_query_state, prep_state.verifier_len) {
+            (Some(_), Some(verifier_len)) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::decode(bytes)?,
+                szk_query_share_opt: Some(SzkQueryShare::<F, SEED_SIZE>::decode_with_param(
+                    &(true, verifier_len),
+                    bytes,
+                )?),
+            }),
+            (None, Some(verifier_len)) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::decode(bytes)?,
+                szk_query_share_opt: Some(SzkQueryShare::<F, SEED_SIZE>::decode_with_param(
+                    &(false, verifier_len),
+                    bytes,
+                )?),
+            }),
+            (None, None) => Ok(MasticPrepareShare {
+                vidpf_proofs: Seed::<SEED_SIZE>::decode(bytes)?,
+                szk_query_share_opt: None,
+            }),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
+        }
+    }
+}
+
+/// Mastic prepare message.
+///
+/// Result of preprocessing the broadcast messages of both aggregators during the
+/// preparation phase.
+pub type MasticPrepareMessage<F, const SEED_SIZE: usize> = Option<SzkVerifier<F, SEED_SIZE>>;
+
+impl<F: FieldElement, const SEED_SIZE: usize> Encode for MasticPrepareMessage<F, SEED_SIZE> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        match self {
+            Some(verifier) => verifier.encode(bytes),
+            None => Ok(()),
+        }
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        match self {
+            Some(verifier) => verifier.encoded_len(),
+            None => Some(0),
+        }
+    }
+}
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareMessage<F, SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match (&prep_state.szk_query_state, prep_state.verifier_len) {
+            (Some(_), Some(verifier_len)) => Ok(Some(
+                SzkVerifier::<F, SEED_SIZE>::decode_with_param(&(true, verifier_len), bytes)?,
+            )),
+            (None, Some(verifier_len)) => Ok(Some(SzkVerifier::<F, SEED_SIZE>::decode_with_param(
+                &(false, verifier_len),
+                bytes,
+            )?)),
+            (None, None) => Ok(None),
+            (Some(_), None) => Err(CodecError::UnexpectedValue),
```

Comment:
Likewise here: this is an issue with the host's state, not an issue with parsing the peer's message.

Reply:
By likewise do you mean we should be able to prevent this situation ever happening? I think that's true, but the match statement here still needs to be exhaustive. I've tried using a different error type.

Reply:
I mean that "unexpected value" is for cases where the peer sent an unexpected value. Here we get this error because the decoding parameter is incorrect. This is our mistake, our peer's.